### PR TITLE
Bug 575057 Mirror files from docker images

### PR DESCRIPTION
**Pull request by cmorty**
This is a major redesign. It reduces the code duplication in `CopyFromContainerCommandHandler` and `ContainerCommandProcess` by introducing the `CopyFromDockerJob`. This `CopyFromDockerJob` can also mirror a folder, creating necessary symlinks and and following symlinks within that folder. This can be used by CDT to copy include paths, which   often contain symlinks to library installation paths outside the   actually included folder.  The code was also extended in a way that should ensure that there are no    stale containers left behind, even if something goes wrong. Even eclipse  crashes the container should self-remove itself after an hour.   Concurrency might be handled in a over-pessimistic way, which might  result in more jobs being blocked then necessary, but should avoid  issues caused by corner cases.

As this is my first bigger contribution (besides some minor bugfixes) I looking forward to feedback. Unfortunately, as this is more or less a rewrite, it wasn't really possible to split this in smaller commits.

There are also two topics that might need fixing:
 * **Copyright**: I copy and pasted quite a lot from the original code (especially `ContainerLauncher.java` and `CopyFromContainerCommandHandler.java`) and then rewrote quite a bit of it. Do I need to copy the copyright from the original files.
 * **Namespace** I placed  `CopyFromDockerJob`  in `org.eclipse.linuxtools.internal.docker.ui.jobs` and  `CloseableContainer` in `org.eclipse.linuxtools.internal.docker.core`. Is this the right place?

### CODE REVIEW

**Comment by jjohnstn** [comment]
For copyright, yes, if you borrow from somewhere you must preserve the copyright.  You should add "and others" if not already specified.  Then, there should be entries under contributor.  Red Hat Inc. would be initial contributor.  You then add yourself or company (if your company owns your copyright) with a brief explanation of what you added (bug number if you want or brief description).  If the file is new, there is one copyright year.  Otherwise, the upper year of the copyright is either added as 2021 or updated to 2021.

I think your namespace selections are fine.  The original copy job had references to PlatformUI so can't be put in core.

**Comment by cmorty** [comment]
> For copyright, yes, if you borrow from somewhere you must preserve the copyright. You should add "and others" if not already specified. Then, there should be entries under contributor. Red Hat Inc. would be initial contributor. You then add yourself or company (if your company owns your copyright) with a brief explanation of what you added (bug number if you want or brief description). If the file is new, there is one copyright year. Otherwise, the upper year of the copyright is either added as 2021 or updated to 2021.
> 
> I think your namespace selections are fine. The original copy job had references to PlatformUI so can't be put in core.

@jjohnstn : Ok. So if I get you right, I merge the copyright headers from the original files into the new one, right? Development was done by writing the new file by copying bits from the old files....

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java [hunk 1]
```diff
@@ -2048,20 +1460,8 @@ private void setConsole(RunConsole cons) {
 	 *
 	 * @since 3.0
 	 */
-	public Set<String> getCopiedVolumes(String connectionName,
```
Comment (jjohnstn) [inline containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java:2051]: Hi Morty.  You cannot remove or modify public APIs.  This would require a major version bump which we don't want to do.  You can add APIs as we are doing a minor release.  So, this API needs to be preserved in the new model.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java [hunk 2]
```diff
@@ -2048,20 +1460,8 @@ private void setConsole(RunConsole cons) {
...
-					.get(connectionName);
-			if (connectionMap != null) {
-				Set<String> imageSet = connectionMap.get(imageName);
-				if (imageSet != null) {
-					copiedSet = imageSet;
-				}
-			}
-		}
-		return copiedSet;
+	public static Set<IPath> getCopiedVolumes(IPath hostdir) {
```
Comment (jjohnstn) [inline containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java:1463]: The comment describing this new method is actually for the previous method.  You need to add a new description.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 3]
```diff
@@ -0,0 +1,1233 @@
...
+
+
+		m_connection = (DockerConnection) connection;
+
+
+		m_containerId = null;
+		if (copyType == CopyType.Container || copyType == CopyType.ContainerMirror) {
+			m_image = null;
+			m_containerId = desc;
+		} else if (copyType == CopyType.ImageMirror || copyType == CopyType.ContainerMirror) {
```
Comment (jjohnstn) [inline containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:239]: I believe this line should be copyType == CopyType.Image || copytType == CopyType.ImageMirror

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 4]
```diff
@@ -0,0 +1,1233 @@
...
+			m_image = desc;
+			m_containerId = null;
+		} else {
+			throw new RuntimeException("Unknow parameter");
```
Comment (jjohnstn) [inline containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:243]: All messages must be translatable.  In addition this message has a misspelling.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 5]
```diff
@@ -0,0 +1,1233 @@
...
+
+		try (InputStream in = m_connection.copyContainer(dcToken, m_containerId, srcdir.toString())) {
+			copyTar(srcdir, hostDir, in, symlinkBacklog, m_monitor);
+		}
+
+		for (SymLink sl : symlinkBacklog) {
+			if (!sl.targetInBase()) {
+				// The link goes outside what was just copied
+				if (linkdepth < MAXLINKDEPTH) {
+					copyFolderFromContainer(dcToken, sl.m_target, sl.hostPath(), linkdepth + 1);
```
Comment (jjohnstn) [inline containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:1219]: The linkdepth never gets incremented anywhere.  Did you mean to have it be: ++linkdepth instead of linkdepth + 1?

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/commands/CopyFromContainerCommandHandler.java [hunk 6]
```diff
@@ -61,10 +42,6 @@
 	private static final String ERROR_COPYING_FROM_CONTAINER_NO_CONNECTION = "command.copyfromcontainer.failure.no_connection"; //$NON-NLS-1$
 	private static final String ERROR_POSSIBLE_CIRCULAR_LINKS = "command.copyfromcontainer.failure.possible_circular_links"; //$NON-NLS-1$
```
Comment (jjohnstn) [inline containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/commands/CopyFromContainerCommandHandler.java:43]: This message currently isn't used.

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Hi Morty.  This is my first review via github.  I prefer gerrit but we have switched.  Anyway, I have added a number of comments for you to address.  As a global comment, all your messages should be translatable (see CommandMessages class usage in CopyFromContainerCommandHandler).  I still need to test this out with the CDT.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 7]
```diff
@@ -0,0 +1,1233 @@
...
+			m_mirror = true;
+			m_targetfolder = copyMap.values().iterator().next();
+		}
+
+
+		getImageId();
+
+		if (m_mirror) {
+			if(!copyMap.values().stream().allMatch(x -> x.equals(m_targetfolder))) {
+				throw new RuntimeException("In mirror-mopde all target paths must point to the same folder");
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:260]: Besides the typo: This specific Exception should only show up, if there is an actual bug in the code using the Job. Thus I don't think it's a good idea to translate it (only extra work). Is it OK if I add a comment clarifying that? I'll go over the other Strings.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:239]: My bad. That was past of the latest refactoring I did. I finally got the tests working. I'll look into writing some.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 8]
```diff
@@ -0,0 +1,1233 @@
...
+		 * This should only be used in mirror-mode and is likely to cause a
+		 * null-ptr-exception. It however does so reliably.
+		 *
+		 * @param filename     The path of the symlink
+		 * @param target       Where the symlink points to (in the source FS)
+		 * @param hostBasePath
+		 * @param isDirectory
+		 */
+		public SymLink(final Path filename, final Path target, Path hostBasePath, final boolean isDirectory) {
+			m_basePath = null; // Will crash in most situations outside mirror mode
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:381]: Bad wording. Changed it to "Can cause a exception in some methods, but these shouldn't be used when using this ctor.". This should never happen. It's more of a note to the reviewer.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 9]
```diff
@@ -0,0 +1,1233 @@
...
+	 * Get a reference to the set of paths that are already mirrored to a targetPath
+	 *
+	 * This can only be used in mirror mode
+	 *
+	 * @param targetPath The path to get the set for
+	 * @return The list of paths mirrored to that folder
+	 */
+	private static HashSet<Path> getCopiedSet(Path targetPath) {
+		synchronized (m_lockObjectManagedInit) {
+			HashSet<Path> copyiedList = m_pathToCopyiedList.get(targetPath);
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:754]: ...fixed all of them. :)

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 10]
```diff
@@ -0,0 +1,1233 @@
...
+		Path currentVolume = null;
+
+		// Windows can only create symlinks to existing files, thus we need a backlog
+		Set<SymLink> symlinkBacklog = new HashSet<>();
+
+		try {
+			initmirror();
+
+			// Check if the volumes have already been copies
+			// To bad we break exceptions using a lambda....
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:1009]: Was a note to self -> I wanted to use a lambda, but exceptions don't seem to get propagated. -> Removed

File: containers/org.eclipse.linuxtools.docker.core/src/org/eclipse/linuxtools/internal/docker/core/DockerConnection.java [hunk 11]
```diff
@@ -2196,44 +2196,26 @@ public void attachContainerOutput(final Closeable token, final String id,
...
-					}
-					for (int i = 0; i < lines.length - 1; ++i) {
-						String line = lines[i];
-						processDirectoryLine(line, path, childList);
-					}
-				}
-				if (!lastLine.isEmpty()) {
-					processDirectoryLine(lastLine, path, childList);
-				}
+				dirListing = pty_stream.readFully();
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.core/src/org/eclipse/linuxtools/internal/docker/core/DockerConnection.java:2214]: The implementation of `readFully()` uses `hasNext()` - so I wasn't too worried:
```java
  public String readFully() {
    final StringBuilder stringBuilder = new StringBuilder();
    while (hasNext()) {
      stringBuilder.append(UTF_8.decode(next().content()));
    }
    return stringBuilder.toString();
  }
```
If there still is a problem it's a timing/concurrency issue. And in that case a different solution is needed (IMO).

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 12]
```diff
@@ -0,0 +1,1233 @@
...
+	private boolean checkIfAlreadyCopyied(Path path)
+	{
+
+		if (!m_mirror) {
+			throw new RuntimeException("This must only called in mirror-mode");
+		}
+
+		synchronized (m_lockObjectManagedInit) {
+			String dockerImageId = getImageId();
+			if (!dockerImageId.equals(m_targetpathToImageIdMap.get(m_targetfolder))) {
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:892]: Nope. Changed the Exception to "Mirroring multiple Images to the same folder is not possible". This is a bug at the call-site.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 5 above)
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:1219]: ? `linkdepth` is a local variable that is incremented with every recursive call. `++linkdepth` would be wrong as it would increment the local recursion level.

**Comment by cmorty** [comment]
@jjohnstn : I tried to address your comments (I hope I caught them all). I put them into separate commits. If you are happy with them I'll squash them into the main commit.

Also: How would you like the review workflow to go? Should I close the conversations that I addressed or should I leave them open at let you close them once you are happy with the solution?

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java]: All strings need to either be translatable or marked with // $NON-NLS-1$  Developers and builds will set a flag to mark lines without either in error so you have to do one or the other.  If the error should never happen, $NON-NLS-1$ is ok but usually an assert is used for such cases.

File: containers/org.eclipse.linuxtools.docker.core/src/org/eclipse/linuxtools/internal/docker/core/DockerConnection.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply containers/org.eclipse.linuxtools.docker.core/src/org/eclipse/linuxtools/internal/docker/core/DockerConnection.java:2214]: I believe it was a case of timing - the data was large and didn't all come back at once.  At the time there wasn't any way to fix it other than processing the lines and allowing more data to come along prior to the next hasNext() call.  Anyway, be aware as this will have to use the old method (even if temporarily) if the same problem still exists with the newer docker-client and dependencies as we will not regress functionality.  You can test by touching a 100000 or 1000000 empty files and copying the whole directory over.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java]: Ok

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java]: Perhaps an assert is in order somewhere to make sure the ctor isn't used incorrectly.

**Review by jjohnstn — COMMENTED** [review]
Copyright fix looks good.

**Review by jjohnstn — COMMENTED** [review]
Fix CopyFromDockerJob copyType comparison commit looks good.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:889]: Fine, but again, you need to translate all your exception messages except for cases of "this should never happen".  In most cases of should never happen unless the caller screws up, there should be an assert to guard against it rather than throwing an exception or else make it impossible for the caller to screw up.

**Review by jjohnstn — COMMENTED** [review]
Fixup typos change looks good.

**Review by jjohnstn — COMMENTED** [review]
Remove unused links looks good.

**Review by jjohnstn — COMMENTED** [review]
format code looks good.

**Review by jjohnstn — COMMENTED** [review]
Make BlockingInputStream static looks good.

**Review by jjohnstn — COMMENTED** [review]
Add missic doc commit is fine, but hopefully will merge with your other patch that renames the getCopyiedPaths method to correct typo.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java [hunk 13]
```diff
@@ -240,7 +237,7 @@ public CopyFromDockerJob(IDockerConnection connection, CopyType copyType, String
 			m_image = desc;
 			m_containerId = null;
 		} else {
-			throw new RuntimeException("Unknow parameter");
```
Comment (jjohnstn) [inline containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java]: Unknown is misspelled.  This kind of error and others should be an assert and throw an AssertionException.  Code will legitimately try/catch RuntimeExceptions for other reasons.  An assert makes sense and you don't have to translate the message plus it is clear it should never happen.

**Review by jjohnstn — CHANGES_REQUESTED** [review]

**Review by jjohnstn — COMMENTED** [review]
Re-Add Set<String> getCopiedVolumes method back commit looks good.

**Comment by jjohnstn** [comment]
Hi Morty, most changes look good, but I had a few comments.  My most serious comment is that you are asserting various things but throwing RuntimeExceptions.  I would like these to be assertions that cause an AssertionError which should never be caught or handled.

Regarding the open conversations, I'll close them along the way.  If you feel a comment is being left open unnecessarily, just point it out to me.

Once you deal with the comments, I would prefer a single commit be created as it conforms more to the review flow I am used to.  We still need to test this, especially with regards to large directories of files to copy.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 5 above)
Comment (jjohnstn) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java:1220]: Ah, ok.

**Comment by cmorty** [comment]
@jjohnstn : Just a quick update: Writing test-cases, I fond some corner-cases I'm currently looking at. The code works, but it could be better.

**Comment by jjohnstn** [comment]
@cmorty Thanks for the update.

**Comment by akurtakov** [comment]
What is the status of this one?

**Comment by cmorty** [comment]
@akurtakov, @jjohnstn :  I added some Testcases. I'd like to eventually also support the disabled test-case, but that goes beyond what is currently supported, and I'd prefer to get the current state merged and do that later.
I'm looking forward to your feedback. If you are happy with the code, I'll rebase it get rid of the `[fixup]` commits.

File: containers/org.eclipse.linuxtools.docker.core/src/org/eclipse/linuxtools/internal/docker/core/DockerConnection.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.core/src/org/eclipse/linuxtools/internal/docker/core/DockerConnection.java:2214]: I did some experiments, even with lots of data (`/include` can become quite big) and never had an issue. Maybe it was a docker-issue that is fixed now.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java]: It's an assertions now. :) - As I'm not a Java-developer, I wasn't really aware of then -> Thanks. :)

**Comment by cmorty** [comment]
@akurtakov, @jjohnstn : What should I do about my Docker-Test-Image, as I have no access to docker [1] (which is totally reasonable)? Should I disable them in the Jenkinsfile or in code?

[1] https://wiki.eclipse.org/Jenkins#I_want_to_build_a_custom_Docker_image_.28with_docker_build.29.2C_but_it_does_not_work._What_should_I_do.3F)

**Comment by jjohnstn** [comment]
>  Hi Morty.  I have been busy with getting JDT ready for migration to github prior to 4.24 M1.  I just tried going to the commits page and copying the suggested: "gh pr checkout 9" but it fails for me with ! [rejected]        refs/pull/9/head -> symlink  (non-fast-forward).  I am not at all familiar with github reviewing as yours is my first.  Do you need to rebase the commits on top of current master for me to check them out successfully??  Regarding your tests that don't work on Jenkins:  they definitely should be in the repo but disabled in Jenkins and it would be nice to have a define I can optionally use to run them from a command line build.

**Comment by cmorty** [comment]
Hi @jjohnstn 

> > Hi Morty.  I have been busy with getting JDT ready for migration to github prior to 4.24 M1.  I just tried going to the commits page and copying the suggested: "gh pr checkout 9" but it fails for me with ! [rejected]        refs/pull/9/head -> symlink  (non-fast-forward).  I am not at all familiar with github reviewing as yours is my first.  Do you need to rebase the commits on top of current master for me to check them out successfully??  

I have no idea. Up to now I used no Github client tools. To me Githuab was a "normal" git remote (with lots of remotes when it comes to PR). Maybe because I've been using github since before all these nice tools, and I never got around to using them. But from the error message you should be able to do `git fetch origin refs/pull/9/head:pr9` and then `git checkout pr9`.

> > Regarding your tests that don't work on Jenkins:  they definitely should be in the repo but disabled in Jenkins and it would be nice to have a define I can optionally use to run them from a command line build.

What is your suggestion to achieve this? Test if there is a docker running and skip otherwise? Tell Jenkins / Maven / ?? to skip these? I'm not familiar with Junit, Mavin nor Jenkins. I'm a C++ developer who knows a bit of Java, so hints are very welcome. ;)

**Comment by jjohnstn** [comment]
> Hi @jjohnstn
> 
> > > Hi Morty.  I have been busy with getting JDT ready for migration to github prior to 4.24 M1.  I just tried going to the commits page and copying the suggested: "gh pr checkout 9" but it fails for me with ! [rejected]        refs/pull/9/head -> symlink  (non-fast-forward).  I am not at all familiar with github reviewing as yours is my first.  Do you need to rebase the commits on top of current master for me to check them out successfully??
> 
> I have no idea. Up to now I used no Github client tools. To me Githuab was a "normal" git remote (with lots of remotes when it comes to PR). Maybe because I've been using github since before all these nice tools, and I never got around to using them. But from the error message you should be able to do `git fetch origin refs/pull/9/head:pr9` and then `git checkout pr9`.
> 
> > > Regarding your tests that don't work on Jenkins:  they definitely should be in the repo but disabled in Jenkins and it would be nice to have a define I can optionally use to run them from a command line build.
> 
> What is your suggestion to achieve this? Test if there is a docker running and skip otherwise? Tell Jenkins / Maven / ?? to skip these? I'm not familiar with Junit, Mavin nor Jenkins. I'm a C++ developer who knows a bit of Java, so hints are very welcome. ;)

I'm not a maven/tycho expert so had to look this up.

The way to exclude tests is to use the <excludes></excludes> option for the tycho-surefire configuration. 

See https://www.eclipse.org/tycho/sitedocs/tycho-surefire-plugin/test-mojo.html

So exclude the test in the current build section that has the tycho-surefire configuration

Then you should be able to add a new profile <profile></profile> with just a tycho-surefire plug-in in it that has a configuration that has an <includes></includes> setting which includes just the one test.  To see a sample profile with a surefire plug-in, look at:

https://wiki.eclipse.org/Tycho/Testing_with_Surefire

To test, one would specify -P<new_profile_name> on the mvn build.  I think this should work.

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Hi Morty.  Looks better.  I tried it out and it copied /usr/include fine.  I also tried updating /usr/include with libpng-devel and it handled the symlinks just fine.  I had a few comments which I have added.  Could you please do a rebase -i and squash all your commits into a single commit that is rebased on a master that is up to date?  I also ran your new test successfully.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java [hunk 14]
```diff
@@ -1826,8 +1258,7 @@ public Process runCommand(String connectionName, String imageName,
 						DataVolumeModel dvm = DataVolumeModel.parseString(dir);
 						switch (dvm.getMountType()) {
 						case HOST_FILE_SYSTEM:
-							String bind = LaunchConfigurationUtils
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java:1829]: Hmm. I did those changes while still working 100% windows. I'll double check.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java [hunk 15]
```diff
@@ -1375,26 +846,10 @@ public Integer getgid() {
...
-									ERROR_NO_CONNECTION_WITH_URI,
-									connectionUri)));
-			return -1;
-		}
 
-		CopyVolumesFromImageJob job = new CopyVolumesFromImageJob(connection,
-				imageName, containerDirs, new ArrayList<String>(), hostDir);
-		job.schedule();
-		return 0;
+		dirMapping.put(new AbstractMap.SimpleEntry<>(connectionUri, imageName), hostDir);
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java:850]: Yes, you are right. I moved it to `fetchContainerDirsJobInt`.

File: containers/org.eclipse.linuxtools.docker.ui.tests/src/org/eclipse/linuxtools/docker/core/CopyFromDockerJobTest.java [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+		localTempWork = Files.createTempDirectory("eclipse-docker-");
+		logger.info("Tempdir: " + localTempWork);
+	}
+
+	@After
+	public void cleanTemp() throws IOException {
+		var paths = Files.walk(localTempWork).collect(Collectors.toList());
+		// Must delete recursvely
+		Collections.reverse(paths);
+		// paths.stream().forEach(p -> p.toFile().delete());
```
[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui.tests/src/org/eclipse/linuxtools/docker/core/CopyFromDockerJobTest.java:69]: No. Thanks for fetching that! I'll try to think of a testcase.

[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui.tests/src/org/eclipse/linuxtools/docker/core/CopyFromDockerJobTest.java]: Ok. That was the testcase - but I also wrote a testcase for the situation where a tag's image changes - and fixed it.

File: containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java:1829]: Sure, do the revert and work on a patch.  It would be good to get the first version of the patch in for M2 which is up-coming.

[the remark this answers is not in the document]
Comment (cmorty) [reply containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/docker/ui/launch/ContainerLauncher.java:1829]: Ok. After a bit of struggling (I thought I was being smart and copied my Linux workspace to Windows) I got it up and working and docker for Windows handles Windows-Paths fine:
grafik

**Comment by cmorty** [comment]
Ok. All the testcases are running on windows. And I compiling from CDT works, too (see https://github.com/eclipse-linuxtools/org.eclipse.linuxtools/pull/9#discussion_r861205217).

**Review by jjohnstn — APPROVED** [review]

**Comment by jjohnstn** [comment]
Thanks Morty.

**Comment by cmorty** [comment]
@jjohnstn : Thanks for baring with me. I'll now improve the docker support in Eclipse. :) I'm also looking forward to podman support. ;-)
